### ISSUE TITLE
Hydration failed because the initial UI does not match what was rendered on the server.

### ISSUE THREAD

**Issue by RyanClementsHax** [rehypejs/rehype-react#43]
### Initial checklist

- [X] I read the [support docs](https://github.com/rehypejs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/rehypejs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/rehypejs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Arehypejs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

rehype-react 7.1.1

### Link to runnable example

https://github.com/RyanClementsHax/ryanclements.dev/tree/server-mismatch-reproduction

### Steps to reproduce

`yarn dev`, open browser to `localhost:3000`, and see the hydration error

This was reproduced using node v19.2.0

The set up I created in this example is processing content on the server, then converting the generated hast to react nodes on the client.

This error seems caused from plugins that generate certain types of nodes in certain orders. In this example, I made a custom plugin that wraps images in `figure`s [here](https://github.com/RyanClementsHax/ryanclements.dev/blob/server-mismatch-reproduction/pages/index.tsx#L19). This will cause the hydration error, but if I wrap the image node in a `picture` tag, everything is fine. I've looked at the generated nodes on the server and the client and didn't notice any differences between them (although I could be wrong). It should be noted that if raw html is parsed from the markdown using `rehype-raw` and such, this all works fine.

Thanks for your hard work and awesome libraries :)

P.S. Let me know if this is not the right repo for this

### Expected behavior

No hydration error occurs

### Actual behavior

Hydration error occurs

### Runtime

Node v17, Other (please specify in steps to reproduce)

### Package manager

yarn 1, Other (please specify in steps to reproduce)

### OS

Linux

### Build and bundle tools

Next.js

**Comment by ChristianMurphy** [comment]
This feels a bit like an [XY problem](https://xyproblem.info/).
Taking a couple steps back.

Your end goal appears to be serverside/statically rendering markdown using Next js, correct?

If so, have you considered the `@next/mdx` integration?
https://mdxjs.com/docs/getting-started/#nextjs

It uses `unified`, `remark`, and `rehype` under the hood, supports `remark` and `rehype` plugins, and can render both `.mdx` files and plain `.md` files.

I'd generally recommend using the pre-built integration/solution, over rolling your own custom static rendering solution.
Unless you have a _very specific_ use case for bespoke rendering, which isn't articulated in your issue yet.

**Comment by RyanClementsHax** [comment]
Thanks for the reply. I've considered those, but wanted to keep the solution lighter and easier for me to understand the inner workings. Also it's been fun learning how to write my own plugins :)

So far everything has been working well until I came by this hydration error which ended up baffling me. To help out, I updated the example repo with a few more additions.

I found out that `rehype-react` will produce this error based on the hast tree its given.

A [`figure` wrapped in a `p`](https://github.com/RyanClementsHax/ryanclements.dev/blob/5a451e2c13695a88f0bb41e7f8ca75b29b9eb413/pages/index.tsx#L98) will cause the issue, but a [`figure` at the root](https://github.com/RyanClementsHax/ryanclements.dev/blob/5a451e2c13695a88f0bb41e7f8ca75b29b9eb413/pages/index.tsx#L71) won't. To check, all you would need to do is replace [useReactFromHast(root)](https://github.com/RyanClementsHax/ryanclements.dev/blob/5a451e2c13695a88f0bb41e7f8ca75b29b9eb413/pages/index.tsx#L58) with the hast you want to check.

This finding doesn't explain the `figure` vs `picture` problem, but it might help?

**Comment by ChristianMurphy** [comment]
The issue is with the validity of the HTML.
`figure` cannot appear inside a `p` tag.

Enter the following static HTML into a browser and inspect the result
```html
<!DOCTYPE html>
<html>
<body>

<p>
  <figure>
    <img src="pic_trulli.jpg" alt="Trulli" style="width:100%">
    <figcaption>Fig.1 - Trulli, Puglia, Italy.</figcaption>
  </figure>
</p>

</body>
</html>
```

note that the browser attempts to fix the invalid HTML.

```html
<!DOCTYPE html>
<html>
<body>

<p></p>
<figure>
  <img src="pic_trulli.jpg" alt="Trulli" style="width:100%">   <figcaption>Fig.1 - Trulli, Puglia, Italy.</figcaption>
</figure>

<p></p>

</body>
</html>
```

that is likely an issue of your server-side rendered producing invalid HTML, the browser fixing it, and the hydrator being unable to reconcile the difference.

**Comment by RyanClementsHax** [comment]
Yup! That ended up being the issue! I didn't know the browser would rewrite the dom if it wasn't valid. I had a hard time catching it because The source page didn't show signs of rewriting and react would overwrite whatever the browser rewrote on first render.

I verified this on the reproduction repo linked.

<img width="337" alt="image" src="https://user-images.githubusercontent.com/20916810/210119233-02255f50-d5a4-478d-890e-782bfdffd4d6.png">

Closing this as this isn't an issue, but a product of me not understanding web standards lol

Thanks for the help :)
